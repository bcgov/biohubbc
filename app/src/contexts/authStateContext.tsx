import useKeycloakWrapper from 'hooks/useKeycloakWrapper';
import React from 'react';

export interface IAuthState {
  ready?: boolean;
}

export const AuthStateContext = React.createContext<IAuthState>({
  ready: false
});

export const AuthStateContextProvider: React.FC = (props) => {
  const keycloakWrapper = useKeycloakWrapper();

  return (
    <AuthStateContext.Provider value={{ ready: keycloakWrapper.keycloak?.authenticated }}>
      {props.children}
    </AuthStateContext.Provider>
  );
};
