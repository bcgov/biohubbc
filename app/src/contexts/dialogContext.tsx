import IconButton from '@material-ui/core/IconButton';
import Snackbar from '@material-ui/core/Snackbar';
import CloseIcon from '@material-ui/icons/Close';
import { Color } from '@material-ui/lab/Alert';
import { ErrorDialog, IErrorDialogProps } from 'components/dialog/ErrorDialog';
import YesNoDialog, { IYesNoDialogProps } from 'components/dialog/YesNoDialog';
import React, { createContext, ReactNode, useState } from 'react';

export interface IDialogContext {
  /**
   * Set the yes no dialog props.
   *
   * Note: Any props that are not provided, will default to whatever value was previously set (or the default value)
   *
   * @memberof IDialogContext
   */
  setYesNoDialog: (props: Partial<IYesNoDialogProps>) => void;
  /**
   * The current yes no dialog props.
   *
   * @type {IYesNoDialogProps}
   * @memberof IDialogContext
   */
  yesNoDialogProps: IYesNoDialogProps;
  /**
   * Set the error dialog props.
   *
   * Note: Any props that are not provided, will default to whatever value was previously set (or the default value)
   *
   * @memberof IDialogContext
   */
  setErrorDialog: (props: Partial<IErrorDialogProps>) => void;
  /**
   * The current error dialog props.
   *
   * @type {IErrorDialogProps}
   * @memberof IDialogContext
   */
  errorDialogProps: IErrorDialogProps;
  /**
   * Set the snackbar props.
   *
   * Note: Any props that are not provided, will default to whatever value was previously set (or the default value)
   *
   * @memberof IDialogContext
   */
  setSnackbar: (props: Partial<ISnackbarProps>) => void;
  /**
   * The current snackbar props.
   *
   * @type {ISnackbarProps}
   * @memberof IDialogContext
   */
  snackbarProps: ISnackbarProps;
}

export interface ISnackbarProps {
  open: boolean;
  onClose: () => void;
  severity?: Color;
  color?: Color;
  snackbarMessage: ReactNode;
}

export const defaultYesNoDialogProps: IYesNoDialogProps = {
  dialogTitle: '',
  dialogText: '',
  open: false,
  onClose: () => {
    // default do nothing
  },
  onNo: () => {
    // default do nothing
  },
  onYes: () => {
    // default do nothing
  }
};

export const defaultErrorDialogProps: IErrorDialogProps = {
  dialogTitle: '',
  dialogText: '',
  open: false,
  onClose: () => {
    // default do nothing
  },
  onOk: () => {
    // default do nothing
  }
};

export const defaultSnackbarProps: ISnackbarProps = {
  snackbarMessage: '',
  open: false,
  onClose: () => {
    // default do nothing
  }
};

export const DialogContext = createContext<IDialogContext>({
  setYesNoDialog: () => {
    // default do nothing
  },
  yesNoDialogProps: defaultYesNoDialogProps,
  setErrorDialog: () => {
    // default do nothing
  },
  errorDialogProps: defaultErrorDialogProps,
  setSnackbar: () => {
    // default do nothing
  },
  snackbarProps: defaultSnackbarProps
});

/**
 * Wraps the provided children in a context that provides various modal dialogs/popups.
 *
 * @param {*} props
 * @return {*}
 */
export const DialogContextProvider: React.FC = (props) => {
  const [yesNoDialogProps, setYesNoDialogProps] = useState<IYesNoDialogProps>(defaultYesNoDialogProps);

  const [errorDialogProps, setErrorDialogProps] = useState<IErrorDialogProps>(defaultErrorDialogProps);

  const [snackbarProps, setSnackbarProps] = useState<ISnackbarProps>(defaultSnackbarProps);

  const setYesNoDialog = function (partialProps: Partial<IYesNoDialogProps>) {
    setYesNoDialogProps({ ...yesNoDialogProps, ...partialProps });
  };

  const setSnackbar = function (partialProps: Partial<ISnackbarProps>) {
    setSnackbarProps({ ...snackbarProps, ...partialProps });
  };

  const setErrorDialog = function (partialProps: Partial<IErrorDialogProps>) {
    setErrorDialogProps({ ...errorDialogProps, ...partialProps });
  };

  return (
    <DialogContext.Provider
      value={{
        setYesNoDialog,
        yesNoDialogProps,
        setErrorDialog,
        errorDialogProps,
        setSnackbar,
        snackbarProps
      }}>
      {props.children}
      <YesNoDialog {...yesNoDialogProps} />
      <ErrorDialog {...errorDialogProps} />
      <Snackbar
        anchorOrigin={{
          vertical: 'bottom',
          horizontal: 'center'
        }}
        open={snackbarProps.open}
        autoHideDuration={6000}
        onClose={() => setSnackbar({ open: false })}
        message={snackbarProps.snackbarMessage}
        action={
          <React.Fragment>
            <IconButton size="small" aria-label="close" color="inherit" onClick={() => setSnackbar({ open: false })}>
              <CloseIcon fontSize="small" />
            </IconButton>
          </React.Fragment>
        }
      />
    </DialogContext.Provider>
  );
};
