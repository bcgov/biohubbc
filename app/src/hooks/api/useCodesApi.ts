import { AxiosInstance } from 'axios';
import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';

/**
 * Returns a set of supported api methods for working with projects.
 *
 * @param {AxiosInstance} axios
 * @return {*} object whose properties are supported api methods.
 */
const useCodesApi = (axios: AxiosInstance) => {
  /**
   * Fetch all code sets.
   *
   * @return {*}  {Promise<IGetAllCodeSetsResponse>}
   */
  const getAllCodeSets = async (): Promise<IGetAllCodeSetsResponse> => {
    const { data } = await axios.get('/api/codes/');

    return data;
  };

  return {
    getAllCodeSets
  };
};

export default useCodesApi;
