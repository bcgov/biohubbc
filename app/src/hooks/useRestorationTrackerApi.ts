import axios from 'axios';
import useAdminApi from './api/useAdminApi';
import useAxios from './api/useAxios';
import useCodesApi from './api/useCodesApi';
import useDraftApi from './api/useDraftApi';
import useExternalApi from './api/useExternalApi';
import useProjectApi, { usePublicProjectApi } from './api/useProjectApi';
import useSearchApi, { usePublicSearchApi } from './api/useSearchApi';
import useUserApi from './api/useUserApi';
import usePermitApi from './api/usePermitApi';
import { useContext } from 'react';
import { ConfigContext } from 'contexts/configContext';

/**
 * Returns a set of supported api methods.
 *
 * @return {*} object whose properties are supported api methods.
 */
export const useRestorationTrackerApi = () => {
  const config = useContext(ConfigContext);
  const apiAxios = useAxios(config?.API_HOST);

  const project = useProjectApi(apiAxios);

  const permit = usePermitApi(apiAxios);

  const search = useSearchApi(apiAxios);

  const codes = useCodesApi(apiAxios);

  const draft = useDraftApi(apiAxios);

  const user = useUserApi(apiAxios);

  const admin = useAdminApi(apiAxios);

  const external = useExternalApi(axios);

  const publicApis = {
    project: usePublicProjectApi(apiAxios),
    search: usePublicSearchApi(apiAxios)
  };

  return {
    project,
    permit,
    search,
    codes,
    draft,
    user,
    admin,
    external,
    public: publicApis
  };
};
