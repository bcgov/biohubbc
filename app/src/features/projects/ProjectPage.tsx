import { Box, CircularProgress, Container, Typography } from '@material-ui/core';
import { useBiohubApi } from 'hooks/useBioHubApi';
import { IProject } from 'interfaces/project-interfaces';
import React, { useEffect, useState } from 'react';
import { useParams } from 'react-router';

// export interface IFormControlsComponentProps {
//   id: number;
//   isDisabled?: boolean;
// }

// const FormControlsComponent: React.FC<IFormControlsComponentProps> = (props) => {
//   const history = useHistory();

//   const navigateToEditProjectPage = (id: number) => {
//     history.push(`/projects/${id}/edit`);
//   };

//   return (
//     <>
//       <Grid container spacing={3}>
//         <Grid container item spacing={3}>
//           <Grid item>
//             <Button variant="text" color="primary" startIcon={<ArrowBack />} onClick={() => history.goBack()}>
//               Back to Projects
//             </Button>
//           </Grid>
//           <Grid item>
//             <Button variant="contained" color="primary" onClick={() => navigateToEditProjectPage(props.id)}>
//               Edit
//             </Button>
//           </Grid>
//         </Grid>
//       </Grid>
//     </>
//   );
// };

/**
 * Page to display a single Project.
 *
 * // TODO WIP
 *
 * @return {*}
 */
const ProjectPage: React.FC = () => {
  const urlParams = useParams();

  const biohubApi = useBiohubApi();

  // TODO this is using IProject in the mean time, but will eventually need something likw IProjectRecord
  const [project, setProject] = useState<IProject | null>(null);

  useEffect(() => {
    const getProject = async () => {
      const projectResponse = await biohubApi.getProject(urlParams['id']);

      if (!projectResponse) {
        // TODO error handling/messaging
        return;
      }

      setProject(projectResponse);
    };

    if (!project) {
      getProject();
    }
  }, [urlParams, biohubApi, project]);

  // const handleChange = () => {};

  // const handleSubmitSuccess = () => {};

  if (!project) {
    return <CircularProgress></CircularProgress>;
  }

  return (
    <Box my={3}>
      <Container>
        <Box mb={3}>
          <Typography variant="h2">{`Project ${project.id}`}</Typography>
        </Box>
        <Box>
          {JSON.stringify(project, null, 2)}
          {/* <FormContainer
            isDisabled={true}
            record={project}
            template={projectTemplate}
            formControlsComponent={(props) => {
              return <FormControlsComponent id={project.id} isDisabled={true} {...props} />;
            }}
            onFormChange={handleChange}
            onFormSubmitSuccess={handleSubmitSuccess}></FormContainer> */}
        </Box>
      </Container>
    </Box>
  );
};

export default ProjectPage;
