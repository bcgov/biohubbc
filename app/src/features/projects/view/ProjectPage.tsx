import Box from '@material-ui/core/Box';
import CircularProgress from '@material-ui/core/CircularProgress';
import Container from '@material-ui/core/Container';
// import List from '@material-ui/core/List';
// import ListItem from '@material-ui/core/ListItem';
// import ListItemIcon from '@material-ui/core/ListItemIcon';
// import ListItemText from '@material-ui/core/ListItemText';
import Paper from '@material-ui/core/Paper';
import ProjectAttachments from 'features/projects/view/ProjectAttachments';
import ProjectDetails from 'features/projects/view/ProjectDetails';
import SurveysListPage from 'features/surveys/list/SurveysListPage';
import { useBiohubApi } from 'hooks/useBioHubApi';
import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';
import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
import React, { useCallback, useEffect, useState } from 'react';
import { useParams } from 'react-router';
import ProjectHeader from './ProjectHeader';
import LocationBoundary from 'features/projects/view/components/LocationBoundary';

/**
 * Page to display a single Project.
 *
 * @return {*}
 */
const ProjectPage: React.FC = () => {
  const urlParams = useParams();
  //const location = useLocation();
  const biohubApi = useBiohubApi();

  const [isLoadingProject, setIsLoadingProject] = useState(false);
  const [projectWithDetails, setProjectWithDetails] = useState<IGetProjectForViewResponse | null>(null);

  const [isLoadingCodes, setIsLoadingCodes] = useState(false);
  const [codes, setCodes] = useState<IGetAllCodeSetsResponse>();

  useEffect(() => {
    const getCodes = async () => {
      const codesResponse = await biohubApi.codes.getAllCodeSets();

      if (!codesResponse) {
        // TODO error handling/messaging
        return;
      }

      setCodes(codesResponse);
    };

    if (!isLoadingCodes && !codes) {
      getCodes();
      setIsLoadingCodes(true);
    }
  }, [urlParams, biohubApi.codes, isLoadingCodes, codes]);

  const getProject = useCallback(async () => {
    const projectWithDetailsResponse = await biohubApi.project.getProjectForView(urlParams['id']);

    if (!projectWithDetailsResponse) {
      // TODO error handling/messaging
      return;
    }

    setProjectWithDetails(projectWithDetailsResponse);
  }, [biohubApi.project, urlParams]);

  useEffect(() => {
    if (!isLoadingProject && !projectWithDetails) {
      getProject();
      setIsLoadingProject(true);
    }
  }, [isLoadingProject, projectWithDetails, getProject]);

  if (!codes || !projectWithDetails) {
    return <CircularProgress className="pageProgress" size={40} />;
  }

  return (
    <>
      <Paper elevation={2} square={true}>
        <ProjectHeader projectWithDetails={projectWithDetails} refresh={getProject} />
      </Paper>

      <Paper elevation={2} square={true}>
        <Container maxWidth="xl">
          <Box display="flex" justifyContent="space-between">
            <Box>
              <ProjectDetails projectForViewData={projectWithDetails} codes={codes} refresh={getProject} />
            </Box>
            <Box>
              <LocationBoundary projectForViewData={projectWithDetails} codes={codes} refresh={getProject} />
            </Box>
          </Box>
          <Box mt={3}>
            <SurveysListPage projectForViewData={projectWithDetails} />
          </Box>
          <Box mt={3}>
            <ProjectAttachments projectForViewData={projectWithDetails} />
          </Box>
        </Container>
      </Paper>
    </>
  );
};

export default ProjectPage;
