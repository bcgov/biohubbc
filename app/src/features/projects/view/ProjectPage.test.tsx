import {
  cleanup,
  //fireEvent,
  render
  // ,
  // waitFor
} from '@testing-library/react';
import { createMemoryHistory } from 'history';
import { useRestorationTrackerApi } from 'hooks/useRestorationTrackerApi';
import { IGetProjectForViewResponse } from 'interfaces/useProjectApi.interface';
import { IGetAllCodeSetsResponse } from 'interfaces/useCodesApi.interface';
import React from 'react';
import { Router } from 'react-router';
//import { getProjectForViewResponse } from 'test-helpers/project-helpers';
import ProjectPage from './ProjectPage';
import { DialogContextProvider } from 'contexts/dialogContext';
//import { SYSTEM_ROLE } from 'constants/roles';
//import { AuthStateContext, IAuthState } from 'contexts/authStateContext';

const history = createMemoryHistory({ initialEntries: ['/admin/projects/1'] });

jest.mock('../../../hooks/useRestorationTrackerApi');
const mockuseRestorationTrackerApi = {
  project: {
    getProjectForView: jest.fn<Promise<IGetProjectForViewResponse>, [number]>(),
    deleteProject: jest.fn(),
    publishProject: jest.fn()
  },
  codes: {
    getAllCodeSets: jest.fn<Promise<IGetAllCodeSetsResponse>, []>()
  }
};

const mockRestorationTrackerApi = ((useRestorationTrackerApi as unknown) as jest.Mock<
  typeof mockuseRestorationTrackerApi
>).mockReturnValue(mockuseRestorationTrackerApi);

// const defaultAuthState = {
//   keycloakWrapper: {
//     keycloak: {
//       authenticated: true
//     },
//     hasLoadedAllUserInfo: true,
//     systemRoles: [SYSTEM_ROLE.SYSTEM_ADMIN] as string[],
//     getUserIdentifier: () => 'testuser',
//     hasAccessRequest: false,
//     hasSystemRole: () => true,
//     getIdentitySource: () => 'idir',
//     username: 'testusername',
//     displayName: 'testdisplayname',
//     email: 'test@email.com',
//     firstName: 'testfirst',
//     lastName: 'testlast',
//     refresh: () => {}
//   }
// };

describe('ProjectPage', () => {
  beforeEach(() => {
    // clear mocks before each test
    mockRestorationTrackerApi().project.deleteProject.mockClear();
    mockRestorationTrackerApi().project.getProjectForView.mockClear();
    mockRestorationTrackerApi().codes.getAllCodeSets.mockClear();
    mockRestorationTrackerApi().project.publishProject.mockClear();
  });

  afterEach(() => {
    cleanup();
  });

  it('renders a spinner if no project is loaded', () => {
    const { asFragment } = render(
      <DialogContextProvider>
        <Router history={history}>
          <ProjectPage />
        </Router>
      </DialogContextProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  // it('renders project page when project is loaded (project is active)', async () => {
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue(getProjectForViewResponse);
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);

  //   const { asFragment, findByText } = render(
  //     <DialogContextProvider>
  //       <Router history={history}>
  //         <ProjectPage />
  //       </Router>
  //     </DialogContextProvider>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   await waitFor(() => {
  //     expect(asFragment()).toMatchSnapshot();
  //   });
  // });

  // it('renders project page when project is loaded (project is completed)', async () => {
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, completion_status: 'Completed' }
  //   });
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);

  //   const { asFragment, findByText } = render(
  //     <DialogContextProvider>
  //       <Router history={history}>
  //         <ProjectPage />
  //       </Router>
  //     </DialogContextProvider>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   await waitFor(() => {
  //     expect(asFragment()).toMatchSnapshot();
  //   });
  // });

  // it('deletes project and takes user to the projects list page when user is a system administrator', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue(getProjectForViewResponse);
  //   mockRestorationTrackerApi().project.deleteProject.mockResolvedValue(true);

  //   const authState = {
  //     keycloakWrapper: {
  //       ...defaultAuthState.keycloakWrapper,
  //       systemRoles: [SYSTEM_ROLE.SYSTEM_ADMIN] as string[],
  //       hasSystemRole: () => true
  //     }
  //   };

  //   const { getByTestId, findByText, getByText } = render(
  //     <AuthStateContext.Provider value={(authState as unknown) as IAuthState}>
  //       <DialogContextProvider>
  //         <Router history={history}>
  //           <ProjectPage />
  //         </Router>
  //       </DialogContextProvider>
  //     </AuthStateContext.Provider>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   fireEvent.click(getByTestId('delete-project-button'));

  //   await waitFor(() => {
  //     expect(
  //       getByText('Are you sure you want to delete this project, its attachments and associated surveys/observations?')
  //     ).toBeInTheDocument();
  //   });

  //   fireEvent.click(getByTestId('yes-button'));

  //   await waitFor(() => {
  //     expect(history.location.pathname).toEqual(`/admin/projects`);
  //   });
  // });

  // it('shows basic error dialog when deleting project call has no response', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue(getProjectForViewResponse);
  //   mockRestorationTrackerApi().project.deleteProject.mockResolvedValue(null);

  //   const authState = {
  //     keycloakWrapper: {
  //       ...defaultAuthState.keycloakWrapper,
  //       systemRoles: [SYSTEM_ROLE.SYSTEM_ADMIN] as string[],
  //       hasSystemRole: () => true
  //     }
  //   };

  //   const { getAllByRole, queryByText, getByText, findByText, getByTestId } = render(
  //     <AuthStateContext.Provider value={(authState as unknown) as IAuthState}>
  //       <DialogContextProvider>
  //         <Router history={history}>
  //           <ProjectPage />
  //         </Router>
  //       </DialogContextProvider>
  //     </AuthStateContext.Provider>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   fireEvent.click(getByTestId('delete-project-button'));

  //   await waitFor(() => {
  //     expect(
  //       getByText('Are you sure you want to delete this project, its attachments and associated surveys/observations?')
  //     ).toBeInTheDocument();
  //   });

  //   fireEvent.click(getByTestId('yes-button'));

  //   await waitFor(() => {
  //     expect(queryByText('Error Deleting Project')).toBeInTheDocument();
  //   });

  //   // Get the backdrop, then get the firstChild because this is where the event listener is attached
  //   //@ts-ignore
  //   fireEvent.click(getAllByRole('presentation')[0].firstChild);

  //   await waitFor(() => {
  //     expect(queryByText('Error Deleting Project')).toBeNull();
  //   });
  // });

  // it('shows error dialog with API error message when deleting project fails', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue(getProjectForViewResponse);
  //   mockRestorationTrackerApi().project.deleteProject = jest.fn(() => Promise.reject(new Error('API Error is Here')));

  //   const authState = {
  //     keycloakWrapper: {
  //       ...defaultAuthState.keycloakWrapper,
  //       systemRoles: [SYSTEM_ROLE.SYSTEM_ADMIN] as string[],
  //       hasSystemRole: () => true
  //     }
  //   };

  //   const { getAllByRole, queryByText, getByText, findByText, getByTestId } = render(
  //     <AuthStateContext.Provider value={(authState as unknown) as IAuthState}>
  //       <DialogContextProvider>
  //         <Router history={history}>
  //           <ProjectPage />
  //         </Router>
  //       </DialogContextProvider>
  //     </AuthStateContext.Provider>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   fireEvent.click(getByTestId('delete-project-button'));

  //   await waitFor(() => {
  //     expect(
  //       getByText('Are you sure you want to delete this project, its attachments and associated surveys/observations?')
  //     ).toBeInTheDocument();
  //   });

  //   fireEvent.click(getByTestId('yes-button'));

  //   await waitFor(() => {
  //     expect(queryByText('API Error is Here')).toBeInTheDocument();
  //   });

  //   // Get the backdrop, then get the firstChild because this is where the event listener is attached
  //   //@ts-ignore
  //   fireEvent.click(getAllByRole('presentation')[0].firstChild);

  //   await waitFor(() => {
  //     expect(queryByText('API Error is Here')).toBeNull();
  //   });
  // });

  // it('sees delete project button as enabled when accessing an unpublished project as a project administrator', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '' }
  //   });
  //   mockRestorationTrackerApi().project.deleteProject.mockResolvedValue(true);

  //   const authState = {
  //     keycloakWrapper: {
  //       ...defaultAuthState.keycloakWrapper,
  //       systemRoles: [SYSTEM_ROLE.PROJECT_CREATOR] as string[],
  //       hasSystemRole: jest.fn().mockReturnValueOnce(true).mockReturnValueOnce(false).mockReturnValueOnce(true)
  //     }
  //   };

  //   const { getByTestId, findByText } = render(
  //     <AuthStateContext.Provider value={(authState as unknown) as IAuthState}>
  //       <DialogContextProvider>
  //         <Router history={history}>
  //           <ProjectPage />
  //         </Router>
  //       </DialogContextProvider>
  //     </AuthStateContext.Provider>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   expect(getByTestId('delete-project-button')).toBeEnabled();
  // });

  // it('sees delete project button as disabled when accessing a published project as a project administrator', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '2021-07-07' }
  //   });
  //   mockRestorationTrackerApi().project.deleteProject.mockResolvedValue(true);

  //   const authState = {
  //     keycloakWrapper: {
  //       ...defaultAuthState.keycloakWrapper,
  //       systemRoles: [SYSTEM_ROLE.PROJECT_CREATOR] as string[],
  //       hasSystemRole: jest.fn().mockReturnValueOnce(true).mockReturnValueOnce(false).mockReturnValueOnce(true)
  //     }
  //   };

  //   const { getByTestId, findByText } = render(
  //     <AuthStateContext.Provider value={(authState as unknown) as IAuthState}>
  //       <DialogContextProvider>
  //         <Router history={history}>
  //           <ProjectPage />
  //         </Router>
  //       </DialogContextProvider>
  //     </AuthStateContext.Provider>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   expect(getByTestId('delete-project-button')).toBeDisabled();
  // });

  // it('does not see the delete button when accessing project as non admin user', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue(getProjectForViewResponse);

  //   const authState = {
  //     keycloakWrapper: {
  //       ...defaultAuthState.keycloakWrapper,
  //       systemRoles: ['Non Admin User'] as string[],
  //       hasSystemRole: () => false
  //     }
  //   };

  //   const { queryByTestId, findByText } = render(
  //     <AuthStateContext.Provider value={(authState as unknown) as IAuthState}>
  //       <DialogContextProvider>
  //         <Router history={history}>
  //           <ProjectPage />
  //         </Router>
  //       </DialogContextProvider>
  //     </AuthStateContext.Provider>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   expect(queryByTestId('delete-project-button')).toBeNull();
  // });

  // it('renders correctly with no end date', async () => {
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: {
  //       ...getProjectForViewResponse.project,
  //       end_date: (null as unknown) as string
  //     }
  //   });
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);

  //   const { asFragment, findByText } = render(
  //     <Router history={history}>
  //       <ProjectPage />
  //     </Router>
  //   );

  //   const projectHeaderText = await findByText('Test Project Name', { selector: 'h1' });
  //   expect(projectHeaderText).toBeVisible();

  //   await waitFor(() => {
  //     expect(asFragment()).toMatchSnapshot();
  //   });
  // });

  // it('publishes and unpublishes a project', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '' }
  //   });
  //   mockRestorationTrackerApi().project.publishProject.mockResolvedValue({ id: 1 });

  //   const { getByTestId, findByText } = render(
  //     <DialogContextProvider>
  //       <Router history={history}>
  //         <ProjectPage />
  //       </Router>
  //     </DialogContextProvider>
  //   );

  //   const publishButtonText1 = await findByText('Publish Project');
  //   expect(publishButtonText1).toBeVisible();

  //   //re-mock response to return the project with a non-null publish date
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '2021-10-10' }
  //   });

  //   fireEvent.click(getByTestId('publish-project-button'));

  //   const unpublishButtonText = await findByText('Unpublish Project');
  //   expect(unpublishButtonText).toBeVisible();

  //   //re-mock response to return the project with a null publish date
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '' }
  //   });

  //   fireEvent.click(getByTestId('publish-project-button'));

  //   const publishButtonText2 = await findByText('Publish Project');
  //   expect(publishButtonText2).toBeVisible();
  // });

  // it('shows API error when fails to publish project', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '' }
  //   });
  //   mockRestorationTrackerApi().project.publishProject = jest.fn(() => Promise.reject(new Error('API Error is Here')));

  //   const { getByTestId, findByText, queryByText, getAllByRole } = render(
  //     <DialogContextProvider>
  //       <Router history={history}>
  //         <ProjectPage />
  //       </Router>
  //     </DialogContextProvider>
  //   );

  //   const publishButtonText1 = await findByText('Publish Project');
  //   expect(publishButtonText1).toBeVisible();

  //   //re-mock response to return the project with a non-null publish date
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '2021-10-10' }
  //   });

  //   fireEvent.click(getByTestId('publish-project-button'));

  //   await waitFor(() => {
  //     expect(queryByText('API Error is Here')).toBeInTheDocument();
  //   });

  //   // Get the backdrop, then get the firstChild because this is where the event listener is attached
  //   //@ts-ignore
  //   fireEvent.click(getAllByRole('presentation')[0].firstChild);

  //   await waitFor(() => {
  //     expect(queryByText('API Error is Here')).toBeNull();
  //   });
  // });

  // it('shows basic error dialog when publish project returns null response', async () => {
  //   mockRestorationTrackerApi().codes.getAllCodeSets.mockResolvedValue({
  //     activity: [{ id: 1, name: 'activity 1' }]
  //   } as any);
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '' }
  //   });
  //   mockRestorationTrackerApi().project.publishProject.mockResolvedValue(null);

  //   const { getByTestId, findByText, queryByText, getAllByRole } = render(
  //     <DialogContextProvider>
  //       <Router history={history}>
  //         <ProjectPage />
  //       </Router>
  //     </DialogContextProvider>
  //   );

  //   const publishButtonText1 = await findByText('Publish Project');
  //   expect(publishButtonText1).toBeVisible();

  //   //re-mock response to return the project with a non-null publish date
  //   mockRestorationTrackerApi().project.getProjectForView.mockResolvedValue({
  //     ...getProjectForViewResponse,
  //     project: { ...getProjectForViewResponse.project, publish_date: '2021-10-10' }
  //   });

  //   fireEvent.click(getByTestId('publish-project-button'));

  //   await waitFor(() => {
  //     expect(queryByText('Error Publishing Project')).toBeInTheDocument();
  //   });

  //   // Get the backdrop, then get the firstChild because this is where the event listener is attached
  //   //@ts-ignore
  //   fireEvent.click(getAllByRole('presentation')[0].firstChild);

  //   await waitFor(() => {
  //     expect(queryByText('Error Publishing Project')).toBeNull();
  //   });
  // });
});
