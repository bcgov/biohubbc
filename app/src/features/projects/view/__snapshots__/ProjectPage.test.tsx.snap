// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectPage renders a spinner if no project is loaded 1`] = `
<DocumentFragment>
  <div
    class="MuiCircularProgress-root pageProgress MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
    role="progressbar"
    style="width: 40px; height: 40px;"
  >
    <svg
      class="MuiCircularProgress-svg"
      viewBox="22 22 44 44"
    >
      <circle
        class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;
